testutil/isstrictnilchecker.go

```diff
@@ -0,0 +1,71 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testutil
+
+import (
+	"fmt"
+	"reflect"
+
+	"gopkg.in/check.v1"
+)
+
+// IsStrictNil checks that the variable is nil typed and nil valued.
+var IsStrictNil = &isStrictNilChecker{
```

Comment:
My thinking was that if the user doesn't know about the issue this is trying to prevent, a longer name won't make it much clearer. At the same time, making it more similar to `IsNil` would make it more likely to be remembered and used. But that's not a very strong opinion, I'm not sure tbh. @pedronis also had thoughts about the name, I'd like to know his preference as well

Reply:
I'd like to question the use-case of this `IsStrictNil` checker. What will this catch that `IsNil` wont catch? AFAICT IsNil will return false in cases of `Interfaces{nil}` (etc..), and that should in theory make a unit test fail anyway if that is unexpected (and force you to be explicit should it be expected/allowed). 

I can see the other way being more dangerous, that you make a `NotNil` check, but you get a `Interface{nil}` and that makes the test pass, where it makes much more sense to do a StrictNotNil (which in theory should be the default we use instead of NotNil).

This may be me missing something (it usually am), but I just don't see the practical use for this.

Reply:
I don't think we should use the concept of nil typed as there is no nil type in the golang spec. The literal nil is untyped and gets typed depending on context. In the context of reflect it seems to be called the zero Value which has an invalid type (IsValid() == false,  Kind() == reflect.Invalid, Type() panics). So maybe:

// IsInterfaceNil checks that the value is a nil interface value.

Other names could be IsZeroValue or IsInvalid but they don't make sense reading only the spec, they make sense only wrt the reflect package.

Reply:
Changed the name and also some of the error messages to be slightly more clear